tests/__init__.py

```diff
@@ -213,11 +214,21 @@ def nocompat_test_suite():
     return result
 
 
+@patch("dulwich.protocol.DEFAULT_GIT_PROTOCOL_VERSION_FETCH", new=0)
+def compat_test_suite_client_git_protocol_v0():
```

Comment:
This should all be down in compat testsuite, rather than complicating things here.

Reply:
I have found a better way to do this. We can subclass all relevant client test classes and mock the default protocol  value on those subclasses. This achieves the same with much less code (and for some reason the code in the version you reviewed didn't even work correctly anymore -- all test runs were using protocol version 2).